core-errors/src/simple.rs

```diff
@@ -16,180 +16,369 @@
 
 //! Simple errors being used for status codes
 
+#[cfg(feature = "codec")]
 use enum_iterator::Sequence;
 
-/// Type that can be encoded and decoded into status code
-pub trait SimpleCodec: sealed::Sealed + Sized {
-    /// Convert type into status code
-    fn into_status_code(self) -> i32;
+#[cfg(feature = "codec")]
+use scale_info::{
+    scale::{self, Decode, Encode},
+    TypeInfo,
+};
 
-    /// Convert status code into self
-    fn from_status_code(status_code: i32) -> Option<Self>;
+#[repr(u8)]
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Default,
+    derive_more::Display,
+    derive_more::From,
+)]
+#[cfg_attr(feature = "codec", derive(Encode, Decode, TypeInfo, Sequence), codec(crate = scale), allow(clippy::unnecessary_cast))]
+/// Enum representing reply code with reason of its creation.
+pub enum ReplyCode {
+    /// Success reply.
+    #[display(fmt = "Success reply sent due to {_0}")]
+    Success(SuccessReason) = 0,
+
+    /// Error reply.
+    #[display(fmt = "Error reply sent due to {_0}")]
+    Error(ErrorReason) = 1,
+
+    /// Unsupported code.
+    /// Variant exists for backward compatibility.
+    #[default]
+    #[display(fmt = "<unsupported reply code>")]
+    Unsupported = 255,
 }
 
-mod sealed {
-    pub trait Sealed {}
+impl ReplyCode {
+    fn discriminant(&self) -> u8 {
+        // SAFETY: Because `Self` is marked `repr(u8)`, its layout is a `repr(C)` `union`
+        // between `repr(C)` structs, each of which has the `u8` discriminant as its first
+        // field, so we can read the discriminant without offsetting the pointer.
+        unsafe { *<*const _>::from(self).cast::<u8>() }
+    }
```

Comment:
Frankly speaking there is no real advantage to using such approach. It may make the code shorter in couple of lines nevertheless there is unnecessary unsafe block.

Reply:
If you want we can change it to just match expression in any further PR.
It may ultra-low performance increase that could be ignored

Reply:
ok